{Doc} Add local context developer guidelines (#13668)
Azure/azure-cli: doc/local_context_guidelines.md

The document provides instructions and guidelines on how to board local context feature for parameters.

## Local context attribute

> Local context attribute is an abstraction of the definition for local context. We use this class to define how the argument support local context.

Reviewer: ```suggestion
Local context attribute is an abstraction of the definition for local context. We use this class to define how the argument supports local context.
```

There are 3 properties of local context attribute.

- **name**

> `name` is the argument name used in local context. Usually just use the name which defined in the function signature, for example `resource_group_name`.

Reviewer: ```suggestion
`name` is the argument name used in local context. Usually we just use the name which defined in the function signature, for example, `resource_group_name`.
```

- **scopes**

> `scopes` is an array of string. It defines where can this local context value been referenced. The value could be a command group or a command, for example `['vm', 'network']`. If one argument can be referenced in all the commands, you can define it as `['all']`. `scopes` is meaningful only when `SET` is in actions.

Reviewer: ```suggestion
`scopes` is an array of string. It defines where this local context value can be referenced. The value could be a command group or a command, for example, `['vm', 'network']`. If one argument can be referenced in all the commands, you can define it as `['all']`. `scopes` is meaningful only when `SET` is in actions.
```

> `scopes` is an array of string. It defines where can this local context value been referenced. The value could be a command group or a command, for example `['vm', 'network']`. If one argument can be referenced in all the commands, you can define it as `['all']`. `scopes` is meaningful only when `SET` is in actions.

Reviewer 2: ```suggestion
`scopes` is an array of strings. It defines where can this local context value been referenced. The value could be a command group or a command, for example `['vm', 'network']`. If one argument can be referenced in all the commands, you can define it as `['all']`. `scopes` is meaningful only when `SET` is in actions.
```

Author: updated

    
- **actions**

> `actions` is an array of [`LocalContextAction`](#Local context action). It defines whether set the command parameter value to local context or get value from local context. We can define both `GET` and `SET` action for one argument.

Reviewer: It is kind of confusing
```
[`LocalContextAction`](#Local context action)
```

Author: This is a link, which will link to the definition for `LocalContextAction`

## Local context action

> Local context action is defined as an enum. The available values are `GET` and `SET`.

Reviewer: ```suggestion
Local context action is defined as an enum type. The available values are `GET` and `SET`.
```

- **GET**

> Used for retrieving value local context.

Reviewer: ```suggestion
Used for retrieving value from local context.
```

- **SET**

Used for saving value to local context.

## Parameters which support `all` scope by default

> We define local context attribute for some command parameters by default. For commands which use them as argument name in the function signature will support local context by default.

Reviewer: Grammar error in the 2nd sentence.

Author: updated

> All these parameters are list here:

Reviewer: ```suggestion
All these parameters are listed here:
```

- *resource_group_name*
```python
    local_context_attribute=LocalContextAttribute(
        name='resource_group_name',
        actions=[LocalContextAction.SET, LocalContextAction.GET],
        scopes=[ALL]
    ))
```

- *vnet_name*
```python
    local_context_attribute=LocalContextAttribute(name='vnet_name', actions=[LocalContextAction.GET])
```

- *subnet*
```python
    local_context_attribute=LocalContextAttribute(name='subnet_name', actions=[LocalContextAction.GET])
```

For example, below is the definition of custom function for `az network vnet create`:
```python
def create_vnet(cmd, resource_group_name, vnet_name, vnet_prefixes='10.0.0.0/16',
                subnet_name=None, subnet_prefix=None, dns_servers=None,
                location=None, tags=None, vm_protection=None, ddos_protection=None,
                ddos_protection_plan=None):
```
As `resource_group_name` is defined in the signature, it will automatically support local context.

## Suggestions

> - Only enable local context feature for resource dependency parameters.

Reviewer: What is "resource dependency parameters"

- Do not define `GET` action for `name` parameter in `create` and `delete` command.
- Define `SET` action only in `create` command.

## Examples

> In order to create a webapp, user need to prepare an appservice plan first. Previously the user need to run below commands to complete this scenario:

Reviewer: ```suggestion
In order to create a webapp, a user needs to prepare an appservice plan first. Previously the user needed to run below commands to complete this scenario:
```

```bash
az group create --name myResourceGroup --location westus
az appservice plan create -g myResourceGroup --name myPlan
az webapp create -g myResourceGroup --plan myPlan --name myWebapp
```

> The user has to input resource group name 3 times and plan name 2 times for this scenario. To enable local context feature, we need to define local context attribute as below:

Reviewer: How about now? How many times are needed when local context is enabled. Try to make a comparison.

```python
# get resource group name from local context by define function argument name as `resource_group_name`
def create_webapp(cmd, resource_group_name, name, plan, runtime=None, startup_file=None,  # pylint: disable=too-many-statements,too-many-branches
                  deployment_container_image_name=None, deployment_source_url=None, deployment_source_branch='master',
                  deployment_local_git=None, docker_registry_server_password=None, docker_registry_server_user=None,
                  multicontainer_config_type=None, multicontainer_config_file=None, tags=None,
                  using_webapp_up=False, language=None):
```

> azure-cli/azure/cli/command_modules/appservice/_params.py
```python
# define SET action for webapp name
# define GET action for plan name
with self.argument_context('webapp create') as c:
    c.argument('name',
                local_context_attribute=LocalContextAttribute(name='web_name', actions=[LocalContextAction.SET], scopes=['webapp']))
    c.argument('plan',
                local_context_attribute=LocalContextAttribute(name='plan_name', actions=[LocalContextAction.GET]))
```

> 

Reviewer: Unnecessary blank lines.

Author: updated
